src/data/mutable_data.rs

```diff
@@ -332,53 +332,45 @@ impl MutableData {
         let mut new_data = self.data.clone();
 
         for (key, val) in insert {
-            if new_data.contains_key(&key) {
-                return Err(ClientError::EntryExists);
+            match new_data.entry(key) {
+                Entry::Occupied(entry) => {
+                    if val != *entry.get() {
+                        return Err(ClientError::EntryExists);
+                    }
+                }
+                Entry::Vacant(entry) => {
+                    let _ = entry.insert(val);
+                }
             }
-            let _ = new_data.insert(key.clone(), val);
         }
 
         for (key, val) in update {
-            if !new_data.contains_key(&key) {
-                return Err(ClientError::NoSuchEntry);
-            }
-            let version_valid = if let Entry::Occupied(mut oe) = new_data.entry(key.clone()) {
-                if val.entry_version != oe.get().entry_version + 1 {
-                    false
-                } else {
-                    let _prev = oe.insert(val);
-                    true
+            match new_data.entry(key) {
+                Entry::Occupied(mut entry) => {
+                    if val.entry_version == entry.get().entry_version + 1 {
```

Comment:
say client sends a sequence of updates: update(v_1), update(v_2), update(v_1)
it will get `invalidSuccessor` on the second `update(v_1)` request.

what's the `idempotent` for this case?